File: tests/src/test/java/messages/ExecutionReportTest.java
package messages;

import java.util.UUID;

import org.agrona.DirectBuffer;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Test;

import agrona.messages.ExecutionReportDecoder;
import agrona.messages.MessageHeaderDecoder;
import sbe.SbeEncoder;

public class ExecutionReportTest {
    private static final Logger log = LogManager.getLogger(ExecutionReportTest.class);

    // Define variables
    double amount;
    double secondaryAmount;
    String secondaryCurrency;
    String currency;
    String side;
    String symbol;
    String deliveryDate;
    String transactTime;
    String quoteID;     
    String quoteRequestID; 
    String dealRequestID;
    String dealID;
    String clientID;
    double fxRate;
    short processed;

    @Test
    public void testSbeExecutionReport() {
        log.info("In testSbeExecutionReport");
        System.out.println("Test Working Directory: " + System.getProperty("user.dir"));

        // Create a random GUID
        UUID uuid = UUID.randomUUID();
        String guid = uuid.toString().substring(0, 8).toUpperCase();

        // Define ExecutionReport data
        amount = 1000;
        secondaryAmount = 810.01;
        secondaryCurrency = "EUR";
        currency = "USD";
        side = "BUY";
        symbol = "EURUSD";
        deliveryDate = "20250101";
        transactTime = "20240101-00:00:00.000";
        quoteID = guid;     
        quoteRequestID = guid;   
        dealRequestID = guid;
        dealID = guid + ".1";
        clientID = "TEST";
        fxRate = 1.23456;
        processed = 0;

        // Initialize the encoder
        SbeEncoder sbeEncoder = new SbeEncoder();

        // Define the data to encode
        DirectBuffer dataBuffer = sbeEncoder.encodeExecutionReport(amount, currency, secondaryAmount, 
            secondaryCurrency, side, symbol, deliveryDate, transactTime, quoteID, quoteRequestID, 
            dealRequestID, dealID, clientID, fxRate, processed);

        // Decode the message
        ExecutionReportDecoder decoder = new ExecutionReportDecoder();
        decoder.wrap(dataBuffer, MessageHeaderDecoder.ENCODED_LENGTH, ExecutionReportDecoder.BLOCK_LENGTH, ExecutionReportDecoder.SCHEMA_VERSION);

        // Verify the decoded message
        assertEquals((long) (amount * 100), decoder.amount().mantissa());
        assertEquals(-2, decoder.amount().exponent());
        assertEquals(currency, decoder.currency());
        assertEquals((long) (secondaryAmount * 100), decoder.secondaryAmount().mantissa());
        assertEquals(-2, decoder.secondaryAmount().exponent());
        assertEquals(secondaryCurrency, decoder.secondaryCurrency());
        assertEquals(side, decoder.side());
        assertEquals(symbol, decoder.symbol());
        assertEquals(deliveryDate, decoder.deliveryDate());
        assertEquals(transactTime, decoder.transactTime());
        assertEquals(quoteRequestID, decoder.quoteRequestID());
        assertEquals(quoteID, decoder.quoteID());
        assertEquals(dealRequestID, decoder.dealRequestID());
        assertEquals(dealID, decoder.dealID());
        assertEquals((long) (fxRate * 100000), decoder.fxRate().mantissa());
        assertEquals(-5, decoder.fxRate().exponent());
        assertEquals(clientID, decoder.clientID());

        log.info("Received and verified deal message");

        log.info("Out testSbeExecutionReport");
    }
}

File: frontend/tests/messages/ExecutionReport.test.js
import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import WebSocketServer from 'jest-websocket-mock';
import { WebSocketProvider, useWebSocket } from 'src/contexts/WebSocketContext.js';

import ExecutionReportEncoder from 'src/aeron/v1/ExecutionReportEncoder.js'
import ExecutionReportDecoder from 'src/aeron/v1/ExecutionReportDecoder.js'
import MessageHeaderEncoder from 'src/aeron/MessageHeaderEncoder.js';

// Mock TextEncoder & TextDecoder
import TextEncoder from '../aeron/TextEncoder.js';
import TextDecoder from '../aeron/TextDecoder.js';
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;

// Mock ExecutionReport data
const data = {
    amount: { mantissa: 1000, exponent: -2 },
    currency: 'USD',
    secondaryAmount: { mantissa: 2000, exponent: -2 },
    secondaryCurrency: 'USD',
    side: 'BUY',
    symbol: 'EURUSD',
    deliveryDate: '20230101',
    transactTime: '20240101-00:00:00.000',
    quoteRequestID: 'test-request',
    quoteID: 'test-quote-id',
    dealRequestID: 'test-request',
    dealID: 'test-deal-id',
    fxRate: { mantissa: 123456, exponent: -5 },
    clientID: 'CLNT'
};

const TestComponent = () => {
    const { sendMessage } = useWebSocket();
    const sendTestMessage = () => {
        // Encode the message
        const buffer = new ArrayBuffer(ExecutionReportEncoder.BLOCK_LENGTH + MessageHeaderEncoder.ENCODED_LENGTH);
        const headerEncoder = new MessageHeaderEncoder();
        const encoder = new ExecutionReportEncoder();
    
        // Encode the data
        encoder.wrapAndApplyHeader(buffer, 0, headerEncoder);
        encoder.encodeamount(data.amount);
        encoder.currency(data.currency);
        encoder.encodesecondaryAmount(data.secondaryAmount);
        encoder.secondaryCurrency(data.currency);
        encoder.side(data.side);
        encoder.symbol(data.symbol);
        encoder.deliveryDate(data.deliveryDate);
        encoder.transactTime(data.transactTime);
        encoder.quoteRequestID(data.quoteRequestID);
        encoder.quoteID(data.quoteID);
        encoder.dealRequestID(data.dealRequestID);
        encoder.dealID(data.dealID);     
        encoder.encodefxRate(data.fxRate);
        encoder.clientID(data.clientID);

        // Send the message on the WebSocket
        sendMessage(buffer);
    };

    return <button onClick={sendTestMessage}>Send ExecutionReport</button>;
};

describe('WebSocket ExecutionReport integration test', () => {
    let server;

    beforeEach(() => {
        server = new WebSocketServer('ws://localhost:8292');
    });

    afterEach(() => {
        WebSocketServer.clean();
    });

    it('receives an ExecutionReport message via WebSocket', async () => {
        const { getByText } = render(
            <WebSocketProvider url="ws://localhost:8292">
                <TestComponent />
            </WebSocketProvider>
        );

        await server.connected;
        await act(async () => {
            fireEvent.click(getByText('Send ExecutionReport'));
        });

        // Allow server to process the message
        await new Promise((resolve) => setTimeout(resolve, 50));

        // Log all server messages
        console.log('Server messages after sending:', server.messages);
    
        // Verify the message was received on the server
        const messages = server.messages.filter(
            (message) => !(typeof message === 'string' && message.includes('Connection established'))
        );
        expect(messages).toHaveLength(1);
    
        // Received message
        const receivedMessage = messages[0];
        console.log('Received message:', receivedMessage);

        // Decode the received message
        const decoder = new ExecutionReportDecoder();
        const buffer = new Uint8Array(receivedMessage).buffer;
        decoder.wrap(buffer, MessageHeaderEncoder.ENCODED_LENGTH);

        console.log('Decoding received message...');

        const decodedData = {
            amount: decoder.decodeamount(),
            currency: decoder.currency(),
            secondaryAmount: decoder.decodesecondaryAmount(),
            secondaryCurrency: decoder.secondaryCurrency(),
            side: decoder.side().replace(/\0/g, ''),
            symbol: decoder.symbol(),
            deliveryDate: decoder.deliveryDate(),
            transactTime: decoder.transactTime(),
            quoteRequestID: decoder.quoteRequestID().replace(/\0/g, ''),
            quoteID: decoder.quoteID().replace(/\0/g, ''),
            dealRequestID: decoder.dealRequestID().replace(/\0/g, ''),
            dealID: decoder.dealID().replace(/\0/g, ''),
            fxRate: decoder.decodefxRate(),
            clientID: decoder.clientID(),
        };

        // Step 6: Assert the decoded data matches the original data
        expect(decodedData).toEqual(data);
    });
});
